### PR TITLE
Introduce support for OS command execution

### PR DESCRIPTION
**Pull request by MadhukaHarith92**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/2852
## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/os/nativeimpl/Exec.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public static Object exec(BMap<BString, Object> command, BMap<BString, Object> env) {
+        List<String> commandList = new ArrayList<>();
+        commandList.add(command.getStringValue(StringUtils.fromString("value")).getValue());
+        String[] arguments = command.getArrayValue(StringUtils.fromString("arguments")).getStringArray();
+        BString[] args = new BString[arguments.length];
+        for (int i = 0; i < arguments.length; i++) {
+            args[i] = StringUtils.fromString(arguments[i]);
+        }
+
+        commandList.addAll(Arrays.stream(args).map(BString::getValue).collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply native/src/main/java/io/ballerina/stdlib/os/nativeimpl/Exec.java]: Simplified the logic in [68f8dc6](https://github.com/ballerina-platform/module-ballerina-os/pull/438/commits/68f8dc6ad5d9a6384f1258a613e72916e4c771cb)

File: docs/proposals/environment-variables.md [hunk 2]
```diff
@@ -2,8 +2,8 @@
 
 _Owners_: @daneshk @MadhukaHarith92  
 _Reviewers_: @daneshk  
-_Created_: 2022/12/18   
-_Updated_: 2022/12/18  
+_Created_: 2022/02/18   
+_Updated_: 2022/02/18
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply docs/proposals/environment-variables.md:6]: No, this is a previous typo.

File: ballerina/os.bal [hunk 3]
```diff
@@ -16,6 +16,19 @@
...
+
+public type Command record {|
+    string value;
+    string[] arguments = [];
+|};
+
+public type EnvProperties record {|
+    never command?;
+    Value...;
+|};
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply ballerina/os.bal]: Fixed in [99e39a9](https://github.com/ballerina-platform/module-ballerina-os/pull/438/commits/99e39a949e29ee508974400cc1f13a6500c717dd)

File: native/src/main/java/io/ballerina/stdlib/os/nativeimpl/Exec.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ */
+public class Exec {
+
+    private Exec() {
+
+    }
+    
+    public static Object exec(BMap<BString, Object> command, BMap<BString, Object> env) {
+        List<String> commandList = new ArrayList<>();
+        commandList.add(command.getStringValue(StringUtils.fromString("value")).getValue());
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply native/src/main/java/io/ballerina/stdlib/os/nativeimpl/Exec.java]: Updated in [c6d1af4](https://github.com/ballerina-platform/module-ballerina-os/pull/438/commits/c6d1af4c39f88339c6d8a2453737c4a3ae78607e)

File: ballerina/process.bal [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    # Returns the standard output as default. Option provided to return standard error by providing file descriptor.
+    # If the process was not finished and exited explicitly by running process.waitForExit(), then process.output() will finish the work and exit and return the output. 
+    # ```ballerina
+    # byte[]|os:Error err = process.output(io:stderr);
+    # ```
+    #
+    # + fileOutputStream - The output stream (`io:stdout` or `io:stderr`) content needs to be returned
+    #
```
Comment (daneshk) [inline ballerina/process.bal:42]:
```suggestion
```

File: docs/proposals/command-execution.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+```ballerina
+# Represents a command.
+#
+# + value - The value of the command
+# + arguments - The arguments of the command
+public type Command record {
+string value;
+string[] arguments;
+};
```
Comment (daneshk) [inline docs/proposals/command-execution.md:28]:
```suggestion
public type Command record {
    string value;
    string[] arguments;
};
```

File: docs/proposals/command-execution.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+public isolated function  exec(Command command, *EnvProperties envProperties) returns Process | Error;
+```
+- The `exec` function can be used to execute OS commands programmatically. Users can pass the command as a raw template. Additionally, users can pass any number of environment properties as key-value pairs that need to be set when executing the command.
+
+```ballerina
+# Environment properties that need to be set when executing the command.
+#
+# + command - command which cannot be a key
+public type EnvProperties record {|
+    never command?;
```
Comment (daneshk) [inline docs/proposals/command-execution.md:48]:
```suggestion
    never command?;
    anydata...
```

File: docs/spec/spec.md [hunk 8]
```diff
@@ -55,3 +55,25 @@ The current user's home directory path can be retrieved using the `os:getUserHom
...
+To retrieve the output of the process, `process.output()` function can be used. This will return the standard output as default. 
+There is an option provided to return standard error by providing file descriptor.
+```ballerina
+byte[]|os:Error err = process.output(io:stderr);
+```
+
+To terminate a process, `process.exit()` function can be used.
+```ballerina
+process.exit();
+```
```
Comment (daneshk) [inline docs/spec/spec.md:79]: Please add `Command`, `EnvProperties`, `Process` definition in the spec.

Comment (daneshk) [reply docs/spec/spec.md:79]: Also update the `content` section

File: docs/proposals/command-execution.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    # ```ballerina
+    # int|os:Error exitCode = process.waitForExit();
+    # ```
+    #
+    # + return - Returns the exit code for the process, or else an `Error` if a failure occurs
+    public isolated function waitForExit() returns int|Error;
+
+    # Provides a stream (to read from), which is made available as the 'standard error', or the 'standard out' of the process.
+    # ```ballerina
+    # stream<string, os:Error?>|os:Error err = process.fread(io:stderr);
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply docs/proposals/command-execution.md]: Fixed in [1aaf2e3](https://github.com/ballerina-platform/module-ballerina-os/pull/438/commits/1aaf2e32f3413cbdc2fba43054673cad9b4ecf26)

File: docs/spec/spec.md (hunk 8 above)
Comment (MadhukaHarith92) [reply docs/spec/spec.md:136]: Fixed in [1aaf2e3](https://github.com/ballerina-platform/module-ballerina-os/pull/438/commits/1aaf2e32f3413cbdc2fba43054673cad9b4ecf26)

File: docs/spec/spec.md [hunk 10]
```diff
@@ -57,16 +58,72 @@ string userHome = os:getUserHome();
...
+    # ```ballerina
+    # process.exit();
+    # ```
+    #
+    public isolated function exit() {
+        return nativeExit(self);
+    }
+}
+```
+
```
Comment (daneshk) [inline docs/spec/spec.md:126]: Can you move at the end. after `output()` and `exit()` functions

**Review by daneshk — APPROVED** [review]
LGTM

**Pull request merged**